### PR TITLE
[Fix] Update confusion_matrix.py

### PR DESCRIPTION
**Pull request by XiandongWang**
## Modification

The confusion_matrix.py is not compatible with the current version of mmseg.

### CODE REVIEW

**Comment by xiexinch** [comment]
Hi @XiandongWang,
Thanks for your contribution! Please fix the lint error, we'll merge it ASAP.

**Comment by OpenMMLab-Assistant-004** [comment]
Hi @XiandongWang,

We'd like to express our appreciation for your valuable contributions to the mmsegmentation. Your efforts have significantly aided in enhancing the project's quality.
It is our pleasure to invite you to join our community thorugh Discord_Special Interest Group (SIG) channel. This is a great place to share your experiences, discuss ideas, and connect with other like-minded people. To become a part of the SIG channel, send a message to the moderator, OpenMMLab, briefly introduce yourself and mention your open-source contributions in the #introductions channel. Our team will gladly facilitate your entry. We eagerly await your presence. Please follow this link to join us: ​https://discord.gg/UjgXkPWNqA.

If you're on WeChat, we'd also love for you to join our community there. Just add our assistant using the WeChat ID: openmmlabwx. When sending the friend request, remember to include the remark "mmsig + Github ID".

Thanks again for your awesome contribution, and we're excited to have you as part of our community!

**Comment by XiandongWang** [comment]
> 

Hi, I've tried a few times, but I keep failing due to this error: 'isort Failed.' Could you give me some advice, please?

**Comment by XiandongWang** [comment]
> Hi @XiandongWang, Thanks for your contribution! Please fix the lint error, we'll merge it ASAP.

Hi, I've tried a few times, but I keep failing due to this error: 'isort Failed.' Could you give me some advice, please?

**Comment by xiexinch** [comment]
> > Hi @XiandongWang, Thanks for your contribution! Please fix the lint error, we'll merge it ASAP.
> 
> Hi, I've tried a few times, but I keep failing due to this error: 'isort Failed.' Could you give me some advice, please?

Hi, I've fixed it. We'll run some tests then this PR will be merged.

**Comment by XiandongWang** [comment]
> > > Hi @XiandongWang, Thanks for your contribution! Please fix the lint error, we'll merge it ASAP.
> > 
> > 
> > Hi, I've tried a few times, but I keep failing due to this error: 'isort Failed.' Could you give me some advice, please?
> 
> Hi, I've fixed it. We'll run some tests then this PR will be merged.

Thanks!

File: tools/analysis_tools/confusion_matrix.py [hunk 1]
```diff
@@ -89,14 +98,16 @@ def plot_confusion_matrix(confusion_matrix,
 
     num_classes = len(labels)
     fig, ax = plt.subplots(
-        figsize=(2 * num_classes, 2 * num_classes * 0.8), dpi=180)
+        figsize=(2 * num_classes, 2 * num_classes * 0.8), dpi=300)
```
Comment (xiexinch) [inline tools/analysis_tools/confusion_matrix.py:101]: Perhaps we should add a parameter to control the `dpi`, it's too large for ADE20K dataset.

File: tools/analysis_tools/confusion_matrix.py [hunk 2]
```diff
@@ -50,15 +54,20 @@ def calculate_confusion_matrix(dataset, results):
         dataset (Dataset): Test or val dataset.
         results (list[ndarray]): A list of segmentation results in each image.
     """
-    n = len(dataset.CLASSES)
+    n = len(dataset.METAINFO['classes'])
     confusion_matrix = np.zeros(shape=[n, n])
     assert len(dataset) == len(results)
-    prog_bar = ProgressBar(len(results))
+    ignore_index = dataset.ignore_index
+    prog_bar = progressbar.ProgressBar(len(results))
```
Comment (xiexinch) [inline tools/analysis_tools/confusion_matrix.py:61]:
```suggestion
    ignore_index = dataset.ignore_index
    reduce_zero_label = dataset.reduce_zero_label
    prog_bar = progressbar.ProgressBar(len(results))
```

File: tools/analysis_tools/confusion_matrix.py [hunk 3]
```diff
@@ -50,15 +54,20 @@ def calculate_confusion_matrix(dataset, results):
...
     for idx, per_img_res in enumerate(results):
         res_segm = per_img_res
-        gt_segm = dataset.get_gt_seg_map_by_idx(idx)
+        gt_segm = \
+            dataset[idx]['data_samples'].gt_sem_seg.data.squeeze().numpy()
+        gt_segm, res_segm = gt_segm.flatten(), res_segm.flatten()
+        to_ignore = gt_segm == ignore_index
```
Comment (xiexinch) [inline tools/analysis_tools/confusion_matrix.py:67]:
```suggestion
        gt_segm = \
            dataset[idx]['data_samples'].gt_sem_seg.data.squeeze().numpy().astype(np.uint8)
        gt_segm, res_segm = gt_segm.flatten(), res_segm.flatten()
        if reduce_zero_label:
            gt_segm = gt_segm - 1
        to_ignore = gt_segm == ignore_index
```

Comment (xiexinch) [reply tools/analysis_tools/confusion_matrix.py:67]: Might add support for `reduze_zero_label`

File: tools/analysis_tools/confusion_matrix.py [hunk 4]
```diff
@@ -5,18 +5,22 @@
 import matplotlib.pyplot as plt
 import numpy as np
 from matplotlib.ticker import MultipleLocator
-from mmengine import Config, DictAction
-from mmengine.utils import ProgressBar, load
+from mmengine.config import Config, DictAction
+from mmengine.registry import init_default_scope
+from mmengine.utils import progressbar
```
Comment (xiexinch) [inline tools/analysis_tools/confusion_matrix.py:10]:
```suggestion
from mmengine.utils import progressbar, mkdir_or_exist
```

Comment (xiexinch) [reply tools/analysis_tools/confusion_matrix.py:10]: Might use `mkdir_or_exist` to ensure `save_dir` exists

**Review by xiexinch — APPROVED** [review]

**Pull request merged**